Run fetching of meta-data of tile files in parallel (for GTC-3035)

danscales opened the pull request:
Run fetching of meta-data of tile files in parallel

When generating the tiles.geojson file, the creation of the RasterSource
objects for each tile involves fetching meta-data from the tile files
from S3, and this can be quite slow when proceeding in serial. So, make
changes to fetch the meta-data in parallel using ProcessPoolExecutor.

For 16 workers, this reduces the fetch time on my local machine by
factor of 5-6, so reduce a nearly two-hour process for processing 2800
tiles to 20 minutes.

This is to help complete GTC-3035.  Creating a tile cache at zoom 14 (and
even zoom 13) takes too long to just create the tiles.geojson, and our
batch jobs never run long enough to complete it, so the tile cache creation
fails.

danscales committed:
Change "docker-compose" to "docker compose"

danscales committed:
Run fetching of meta-data of tile files in parallel

When generating the tiles.geojson file, the creation of the RasterSource
objects for each tile involves fetching meta-data from the tile files
from S3, and this can be quite slow when proceeding in serial. So, make
changes to fetch the meta-data in parallel using ProcessPoolExecutor.

For 16 workers, this reduces the fetch time on my local machine by
factor of 5-6, so reduce a nearly two-hour process for processing 2800
tiles to 20 minutes.

dmannarino commented on gfw_pixetl/pixetl_prep.py:
```diff
@@ -3,35 +3,66 @@
+    future_tiles = {}
+    tiles: List[DummyTile] = list()
+
+    with ProcessPoolExecutor(max_workers=16) as executor:
```
Rather than hard-code to 16, change this to use GLOBALS.num_processes as here:
https://github.com/wri/gfw_pixetl/blob/master/gfw_pixetl/pipes/raster_pipe.py#L60
This is meant to scale with the instance size.

danscales replied:
OK, I changed it to min(16, GLOBAL.num_processes), because I don't think we want to overwhelm the single S3 bucket with requests from more than 16 processes (eventually it starts rate-limiting, I think?), since we're not doing anything in these processes other than the S3 meta-data read.

dmannarino replied:
👍

dmannarino commented:
Just one fix needed, otherwise looks great!

danscales committed:
Use GLOBALS.num_processes for parallelism, up to max of 16

I don't think we want to go beyond 16, because we don't want to overload
the S3 bucket we're reading from.

dmannarino reviewed (approved):
Looks good! :)

danscales committed:
Add print() statements.

danscales commented:
This is no longer needed, because Daniel moved the needed pixetl code for the resample.sh script (including geojson creation), which is used when creating a raster tile cache, directly into the data API and that solved the problem.
